# Quickstart

Alice, Bob and Eve work at Critical Labs. The threesome are avid ping pong players. And they're all incredibly competitive. One day, Eve has an idea... in an attempt to find out who actually is <i>The Best Ping Pong Player in the Office</i> she's offered to design a system to keep track of everyone's [Elo rankings](https://en.wikipedia.org/wiki/Elo_rating_system). 

Alice and Bob are cool with her code, so they just let her keep a ledger:

In [1]:
def calculate_elo(winner, loser, k=32):
    expected_probability_winner = 1 / ( 1 + 10**((loser - winner)/400) )
    expected_probability_loser = 1 - expected_probability_winner
    winner = round(winner + k*(1 - expected_probability_winner))
    loser = round(loser + k*(0 - expected_probability_loser))
    return winner, loser

def update_elo(winner, loser, rankings):
    rankings[winner], rankings[loser] = calculate_elo(
    rankings[winner], rankings[loser])
    return rankings

Eve starts each coworker at an Elo ranking of 1000:

In [2]:
pong = {'alice': 1000, 'bob': 1000, 'eve': 1000}

In the first officially tracked game, Alice beats Bob. Accordingly, Eve updates their Elo scores to reflect the outcome: 

In [3]:
update_elo(winner='alice', loser='bob', rankings=pong)

{'alice': 1016, 'bob': 984, 'eve': 1000}

On the same day Eve then beats Alice:

In [4]:
update_elo(winner='eve', loser='alice', rankings=pong)

{'alice': 999, 'bob': 984, 'eve': 1017}

And then Bob beats Alice, and crushes Eve shortly afterward:

In [5]:
update_elo(winner='bob', loser='alice', rankings=pong)
update_elo(winner='bob', loser='eve', rankings=pong)

{'alice': 982, 'bob': 1018, 'eve': 1000}

Just before she goes home, Eve pickles the rankings:

In [6]:
import pickle
with open('pong.pkl', 'wb') as f:
    pickle.dump(pong, f)
del pong

At lunch the next day Alice beats Eve, Bob beats Eve, and then Alice beats Eve, again:

In [7]:
with open('pong.pkl', 'rb') as f:
    pong = pickle.load(f)

In [8]:
update_elo(winner='alice', loser='eve', rankings=pong)
update_elo(winner='bob', loser='eve', rankings=pong)
update_elo(winner='alice', loser='eve', rankings=pong)

{'alice': 1014, 'bob': 1032, 'eve': 954}

Eve is pretty upset that she's at the bottom of the totem pole... so she decides to cook the books once Bob and Alice leave work:

In [9]:
pong = {'alice': 1014, 'bob': 1032, 'eve': 954}
pong['eve'] += 40
pong['alice'] -= 40
pong

{'alice': 974, 'bob': 1032, 'eve': 994}

In [10]:
with open('pong.pkl', 'wb') as f:
    pickle.dump(pong, f)
del pong

On *Day 3* of this little Elo experiment, Alice beats Eve, but then Eve beats Bob:

In [11]:
with open('pong.pkl', 'rb') as f:
    pong = pickle.load(f)

In [12]:
update_elo(winner='alice', loser='eve', rankings=pong)
update_elo(winner='eve', loser='bob', rankings=pong)

{'alice': 991, 'bob': 1013, 'eve': 996}

Alice asks Eve to show her the rankings to see how things are going:

In [13]:
pong

{'alice': 991, 'bob': 1013, 'eve': 996}

Alice: "How the hell are you higher than me!"

Eve: "Idk, I guess I'm just better... get over it"

Alice: "You are NOT! I beat you at least three or four times"

Eve: "Yeah and then I just beat Bob a couple times, that's how it works"

Alice: "🐍"

In [14]:
with open('pong.pkl', 'wb') as f:
    pickle.dump(pong, f)
del pong

After her confrontation with Eve, Alice does a bit of research and finds this cool new Python package called [sausagelink](https://github.com/maxhumber/sausagelink) which she thinks might be just what the doctor order.

*Day 4*, Alice strolls into into work and tells Eve that she would prefer it if she kept track of everything with `sausagelink`. Eve is hesitant, but eventually acquiesces.

Eve fires up her ledger:

In [15]:
with open('pong.pkl', 'rb') as f:
    pong = pickle.load(f)

And then installs the package:

In [16]:
#! pip install sausage link

Not trusting Eve, Alice forces her immediately add the rankings as a `Sausage` to a `Link`:

In [17]:
from sausagelink import Sausage, Link
pong = Link(pong)

> Note: `Link` accepts raw data or `Sausage` data objects and handles them accordingly.

In [18]:
pong

[⎨0⎬]

In [19]:
pong[0] # heeehehheee looks like a sausage

⎨0⎬

In [20]:
[method for method in dir(pong[0]) if not method.startswith('__')]

['data', 'grill', 'index', 'previous_sizzle', 'rancid', 'sizzle', 'timestamp']

In [21]:
pong[0].data

{'alice': 991, 'bob': 1013, 'eve': 996}

Eve: "There. Are you happy!"

Alice: "Actually... yes. Yes I am. Thank you. \*cough\* <sub>cheater</sub> \*cough\* \*cough\*"

Eve: "What did you say?"

Alice: "Nothing"

At lunch, Alice beats Eve, Eve beats Bob, but then Bob beats Eve:

In [22]:
from copy import deepcopy
day_4 = deepcopy(pong[0].data)

In [23]:
day_4

{'alice': 991, 'bob': 1013, 'eve': 996}

In [24]:
update_elo(winner='alice', loser='eve', rankings=day_4)
update_elo(winner='eve', loser='bob', rankings=day_4)
update_elo(winner='bob', loser='eve', rankings=day_4)

{'alice': 1007, 'bob': 1011, 'eve': 982}

In [25]:
pong.append(day_4)

In [26]:
pong

[⎨0⎬, ⎨1⎬]

In [27]:
[day.data for day in pong]

[{'alice': 991, 'bob': 1013, 'eve': 996},
 {'alice': 1007, 'bob': 1011, 'eve': 982}]

Alice still doesn't trust Eve so she has her prove that she recorded everything properly. 
Eve runs the `rancid` method on the entire `Link` and then on the first `Sausage`:

In [28]:
pong.rancid()

[]

In [29]:
pong[0].rancid()

False

The `Link` comes back empty and the Sausage comes back `False` so Alice is satisfied.

Just before she heads home, Eve saves the ledger to a new sausagelink-compatible file:

In [30]:
pong.refrigerate('pong.sl')
del pong

On *Day 5*, Eve just gets absolutely demo-ed:

In [31]:
pong = Link('pong.sl')

In [32]:
day_5 = deepcopy(pong[-1].data)
day_5

{'alice': 1007, 'bob': 1011, 'eve': 982}

In [33]:
update_elo(winner='alice', loser='eve', rankings=day_5)
update_elo(winner='bob', loser='eve', rankings=day_5)
update_elo(winner='bob', loser='eve', rankings=day_5)
update_elo(winner='alice', loser='bob', rankings=day_5)

{'alice': 1039, 'bob': 1021, 'eve': 940}

Alice forces Eve to add the results to the `Link`:

In [34]:
pong.append(day_5)
[day.data for day in pong]

[{'alice': 991, 'bob': 1013, 'eve': 996},
 {'alice': 1007, 'bob': 1011, 'eve': 982},
 {'alice': 1039, 'bob': 1021, 'eve': 940}]

Right after work Eve tries to cook the books, again:

In [35]:
pong[-1].data

{'alice': 1039, 'bob': 1021, 'eve': 940}

In [36]:
pong[-1].data = {'alice': 1019, 'bob': 1021, 'eve': 960}

In [37]:
pong[-1].data

{'alice': 1019, 'bob': 1021, 'eve': 960}

In [38]:
pong.refrigerate('pong.sl')
del pong

*Day 6*: Alice grabs a copy of `pong.sl` and inspects the data herself:

In [39]:
pong = Link('pong.sl')

In [40]:
pong[-1].data

{'alice': 1019, 'bob': 1021, 'eve': 960}

Alice: "What the hell. I do not have 1019. I distinctly remember I had 1039!"

Eve: "No. You had 1019!"

Bob: "Actually, Eve, I'm pretty sure Alice had 1039 and you had 940"

Alice: "Let's check..."

In [41]:
pong[-1].rancid()

True

Alice: "See! I knew it, you screwed with the data after it was inserted"

In [42]:
data = pong[-1]
print(data.grill())
print(data.sizzle)

f8c6bf94cbbf91139aa6dcc40ef82cc5864e907103cd75b55de1deb6b9533792
d50944a78099c354afbe095dbf5a6e80b567d940e6e35d3a9d40508c3421eab6


Alice: "The SHA256 digests don't match! That's why `rancid` is returning True!"

Eve: "Shit..."

In [43]:
import os
os.remove('pong.pkl')
os.remove('pong.sl')